File: audit.py
#!/usr/bin/env python3

import json
import sys
import urllib.request
import urllib.error

URL = "https://ofuofhojhbcrcahuotya.supabase.co"
ANON_KEY = "sb_publishable_Xeg2L1pOa5YK6RjTUuYvNA_7B0tfRr5"

def query(table, params=""):
    url = f"{URL}/rest/v1/{table}?{params}"
    headers = {
        "apikey": ANON_KEY,
        "Authorization": f"Bearer {ANON_KEY}",
    }

    req = urllib.request.Request(url, headers=headers)
    try:
        with urllib.request.urlopen(req) as response:
            return json.loads(response.read().decode())
    except urllib.error.HTTPError as e:
        print(f"HTTP Error {e.code}: {e.read().decode()}", file=sys.stderr)
        return []

def main():
    print("=" * 80)
    print("WASHINGTON ALPINE/MOUNTAINEERING DATABASE AUDIT")
    print("=" * 80)
    print()

    # Step 1: Fetch all areas
    print("Step 1: Fetching all areas...")
    areas = query("areas", "select=*&order=name")
    print(f"  Total areas in database: {len(areas)}")
    print()

    # Step 2: Find Washington areas
    print("Step 2: Finding Washington state and peaks...")
    wa = next((a for a in areas if a["name"] == "Washington"), None)
    if not wa:
        print("ERROR: Washington state not found!")
        return

    print(f"  Found Washington (id: {wa['id']})")

    # Find all areas with "wa" in path (descendants of Washington)
    wa_areas = [a for a in areas if a.get("path", "").startswith("wa")]
    wa_peaks = [a for a in wa_areas if a.get("route_count", 0) > 0 and a.get("area_type") != "state"]

    print(f"  Found {len(wa_peaks)} peaks with routes in Washington")
    print()

    # Step 3: Fetch all routes
    print("Step 3: Fetching all routes...")
    routes = query("routes", "select=*")
    print(f"  Total routes in database: {len(routes)}")

    # Filter routes by WA peaks
    wa_peak_ids = {p["id"] for p in wa_peaks}
    wa_routes = [r for r in routes if r.get("area_id") in wa_peak_ids]
    print(f"  Routes in WA alpine/mountaineering peaks: {len(wa_routes)}")
    print()

    # Step 4: Route counts by peak
    print("Step 4: ROUTE COUNTS BY PEAK (Top 30)")
    print("-" * 80)

    routes_by_peak = {}
    for route in wa_routes:
        area_id = route.get("area_id")
        if area_id not in routes_by_peak:
            peak = next((p for p in wa_peaks if p["id"] == area_id), None)
            routes_by_peak[area_id] = {
                "name": peak["name"] if peak else "UNKNOWN",
                "routes": []
            }
        routes_by_peak[area_id]["routes"].append(route)

    sorted_peaks = sorted(
        [(id, data) for id, data in routes_by_peak.items()],
        key=lambda x: len(x[1]["routes"]),
        reverse=True
    )

    print(f"{'Peak Name':<40} {'Routes':<10} Area ID")
    print("-" * 80)
    for i, (peak_id, data) in enumerate(sorted_peaks[:30]):
        print(f"{data['name'][:39]:<40} {len(data['routes']):<10} {peak_id}")
    print()

    # Step 5: Check specific peaks
    print("Step 5: SPECIFIC PEAK VERIFICATION")
    print("-" * 80)

    adams_peak = next((p for p in wa_peaks if "Mount Adams" in p.get("name", "")), None)
    shuksan_peak = next((p for p in wa_peaks if "Mount Shuksan" in p.get("name", "")), None)

    if adams_peak:
        adams_routes = [r for r in wa_routes if r.get("area_id") == adams_peak["id"]]
        status = "OK" if len(adams_routes) == 7 else "MISMATCH"
        print(f"Mount Adams: {len(adams_routes)} routes (expected 7) [{status}]")
        for r in adams_routes:
            print(f"  - {r.get('name')} ({r.get('grade', 'NO GRADE')}) [{r.get('discipline')}]")
    else:
        print("Mount Adams: NOT FOUND")
    print()

    if shuksan_peak:
        shuksan_routes = [r for r in wa_routes if r.get("area_id") == shuksan_peak["id"]]
        status = "OK" if len(shuksan_routes) == 10 else "MISMATCH"
        print(f"Mount Shuksan: {len(shuksan_routes)} routes (expected 10) [{status}]")
        for r in shuksan_routes:
            print(f"  - {r.get('name')} ({r.get('grade', 'NO GRADE')}) [{r.get('discipline')}]")
    else:
        print("Mount Shuksan: NOT FOUND")
    print()

    # Step 6: Data quality check
    print("Step 6: DATA COMPLETENESS CHECK")
    print("-" * 80)

    VALID_DISCIPLINES = {"alpine", "mountaineering", "rock", "ice", "mixed", "bouldering"}

    missing_name = [r for r in wa_routes if not r.get("name") or not r["name"].strip()]
    missing_grade = [r for r in wa_routes if not r.get("grade") or not r["grade"].strip()]
    missing_discipline = [r for r in wa_routes if not r.get("discipline")]
    missing_area_id = [r for r in wa_routes if not r.get("area_id")]
    invalid_discipline = [r for r in wa_routes if r.get("discipline") and r["discipline"].lower() not in VALID_DISCIPLINES]

    # Check for duplicates
    duplicates = {}
    for r in wa_routes:
        key = (r.get("area_id"), r.get("name"))
        if key in duplicates:
            duplicates[key].append(r)
        else:
            duplicates[key] = [r]
    dup_list = [v for v in duplicates.values() if len(v) > 1]

    print(f"Total routes checked: {len(wa_routes)}")
    print(f"Routes missing name: {len(missing_name)}")
    print(f"Routes missing grade: {len(missing_grade)}")
    print(f"Routes missing discipline: {len(missing_discipline)}")
    print(f"Routes missing area_id: {len(missing_area_id)}")
    print(f"Routes with invalid discipline: {len(invalid_discipline)}")
    print(f"Duplicate route names: {len(dup_list)}")
    print()

    # Report issues
    if dup_list:
        print("*** DUPLICATE ROUTES FOUND ***")
        for dup_routes in dup_list:
            names = [r.get("name") for r in dup_routes]
            area_id = dup_routes[0].get("area_id")
            ids = [r.get("id") for r in dup_routes]
            print(f"  - '{names[0]}' in area {area_id} (IDs: {', '.join(ids)})")
        print()

    if missing_grade:
        print("*** ROUTES MISSING GRADE ***")
        for r in missing_grade[:5]:
            print(f"  - '{r.get('name')}' (area: {r.get('area_id')})")
        if len(missing_grade) > 5:
            print(f"  ... and {len(missing_grade) - 5} more")
        print()

    if missing_discipline:
        print("*** ROUTES MISSING DISCIPLINE ***")
        for r in missing_discipline[:5]:
            print(f"  - '{r.get('name')}' (area: {r.get('area_id')})")
        if len(missing_discipline) > 5:
            print(f"  ... and {len(missing_discipline) - 5} more")
        print()

    if invalid_discipline:
        print("*** ROUTES WITH INVALID DISCIPLINE ***")
        for r in invalid_discipline[:10]:
            print(f"  - '{r.get('name')}': discipline='{r.get('discipline')}' (area: {r.get('area_id')})")
        if len(invalid_discipline) > 10:
            print(f"  ... and {len(invalid_discipline) - 10} more")
        print()

    # Step 7: Discipline breakdown
    print("Step 7: DISCIPLINE BREAKDOWN")
    print("-" * 80)

    disciplines = {}
    for r in wa_routes:
        d = r.get("discipline") or "NONE"
        disciplines[d] = disciplines.get(d, 0) + 1

    for d, count in sorted(disciplines.items(), key=lambda x: -x[1]):
        print(f"{d:<20}: {count}")
    print()

    # Step 8: Grade distribution
    print("Step 8: GRADE DISTRIBUTION (Top 20)")
    print("-" * 80)

    grades = {}
    for r in wa_routes:
        g = r.get("grade") or "NONE"
        grades[g] = grades.get(g, 0) + 1

    top_grades = sorted(grades.items(), key=lambda x: -x[1])[:20]
    for g, count in top_grades:
        print(f"{g:<20}: {count}")
    print()

    # Step 9: Overall health assessment
    print("Step 9: OVERALL HEALTH ASSESSMENT")
    print("-" * 80)

    total_issues = len(missing_name) + len(missing_grade) + len(missing_discipline) + len(dup_list) + len(invalid_discipline)
    issue_percent = (total_issues / len(wa_routes) * 100) if wa_routes else 0

    if issue_percent <= 2:
        health_status = "EXCELLENT"
    elif issue_percent <= 5:
        health_status = "GOOD"
    elif issue_percent <= 10:
        health_status = "FAIR"
    else:
        health_status = "POOR"

    print(f"Total routes: {len(wa_routes)}")
    print(f"Total peaks: {len(sorted_peaks)}")
    print(f"Issues found: {total_issues} ({issue_percent:.2f}%)")
    print(f"Health Status: {health_status}")
    print()

    # Step 10: Recommendations
    if total_issues > 0:
        print("RECOMMENDED FIXES:")
        print("-" * 80)
        if missing_grade:
            print(f"1. Add missing grades to {len(missing_grade)} routes")
        if missing_discipline:
            print(f"2. Add missing disciplines to {len(missing_discipline)} routes")
        if invalid_discipline:
            print(f"3. Fix invalid disciplines in {len(invalid_discipline)} routes")
        if dup_list:
            print(f"4. Remove {len(dup_list)} duplicate route(s)")
    else:
        print("NO ISSUES FOUND - Database is in excellent condition!")

    print()
    print("=" * 80)
    print("AUDIT COMPLETE")
    print("=" * 80)

if __name__ == "__main__":
    main()

File: audit.mjs
#!/usr/bin/env node

const URL = "https://ofuofhojhbcrcahuotya.supabase.co";
const ANON_KEY = "sb_publishable_Xeg2L1pOa5YK6RjTUuYvNA_7B0tfRr5";

async function query(table, options = {}) {
  const { select = "*", eq = null, is = null, order = null } = options;

  let url = `${URL}/rest/v1/${table}?select=${encodeURIComponent(select)}`;

  if (eq) {
    for (const [col, val] of Object.entries(eq)) {
      url += `&${col}=eq.${encodeURIComponent(val)}`;
    }
  }

  if (is) {
    for (const [col, val] of Object.entries(is)) {
      url += `&${col}=is.${val}`;
    }
  }

  if (order) {
    url += `&order=${order}`;
  }

  const response = await fetch(url, {
    headers: {
      Authorization: `Bearer ${ANON_KEY}`,
      apikey: ANON_KEY,
      "Content-Type": "application/json",
    },
  });

  if (!response.ok) {
    throw new Error(`Query failed: ${response.status} ${await response.text()}`);
  }

  return response.json();
}

async function rpc(functionName, args = {}) {
  const response = await fetch(`${URL}/rest/v1/rpc/${functionName}`, {
    method: "POST",
    headers: {
      Authorization: `Bearer ${ANON_KEY}`,
      apikey: ANON_KEY,
      "Content-Type": "application/json",
    },
    body: JSON.stringify(args),
  });

  if (!response.ok) {
    throw new Error(
      `RPC failed: ${response.status} ${await response.text()}`
    );
  }

  return response.json();
}

async function audit() {
  console.log("=".repeat(80));
  console.log("WASHINGTON ALPINE/MOUNTAINEERING DATABASE AUDIT");
  console.log("=".repeat(80));
  console.log();

  try {
    // 1. Get Washington state area
    console.log("Step 1: Fetching Washington state area...");
    const waAreas = await query("areas", {
      eq: { name: "Washington" },
    });

    const wa = waAreas[0];
    if (!wa) {
      console.error("ERROR: Washington state not found!");
      return;
    }
    console.log(`Found Washington (id: ${wa.id})`);
    console.log();

    // 2. Get all areas in WA
    console.log("Step 2: Fetching all areas in Washington...");
    const allAreas = await query("areas", {
      select: "*",
      order: "name",
    });

    // Build area tree - find all areas with WA as ancestor
    const buildAncestors = (areaId) => {
      const area = allAreas.find((a) => a.id === areaId);
      if (!area || !area.parent_id) return [area];
      return [...buildAncestors(area.parent_id), area];
    };

    const peaksWithRoutes = allAreas.filter((a) => {
      if (a.route_count === 0 || a.area_type === "state") return false;
      const ancestors = buildAncestors(a.id);
      return ancestors.some((anc) => anc.id === wa.id);
    });

    console.log(`Found ${peaksWithRoutes.length} peaks with routes in WA`);
    console.log();

    // 3. Fetch all routes
    console.log("Step 3: Fetching all routes in Washington peaks...");
    const allRoutes = await query("routes", {
      select: "*",
    });

    const waRoutes = allRoutes.filter((r) => {
      return peaksWithRoutes.some((p) => p.id === r.area_id);
    });

    console.log(
      `Total routes found in WA alpine/mountaineering peaks: ${waRoutes.length}`
    );
    console.log();

    // 4. Route counts by peak
    console.log("Step 4: ROUTE COUNTS BY PEAK");
    console.log("-".repeat(80));
    const routesByPeak = {};
    waRoutes.forEach((r) => {
      if (!routesByPeak[r.area_id]) {
        const peak = peaksWithRoutes.find((p) => p.id === r.area_id);
        routesByPeak[r.area_id] = {
          name: peak?.name || "UNKNOWN",
          routes: [],
        };
      }
      routesByPeak[r.area_id].routes.push(r);
    });

    const sortedPeaks = Object.entries(routesByPeak)
      .map(([id, data]) => ({ id, ...data, count: data.routes.length }))
      .sort((a, b) => b.count - a.count);

    console.log("Peak Name".padEnd(40) + "Routes".padEnd(10) + "Area ID");
    console.log("-".repeat(80));
    sortedPeaks.forEach((p) => {
      console.log(
        p.name.substring(0, 39).padEnd(40) +
          p.count.toString().padEnd(10) +
          p.id
      );
    });
    console.log();

    // 5. Check specific peaks
    console.log("Step 5: SPECIFIC PEAK VERIFICATION");
    console.log("-".repeat(80));
    const adamsPeak = peaksWithRoutes.find((p) =>
      p.name.includes("Mount Adams")
    );
    const shuksanPeak = peaksWithRoutes.find((p) =>
      p.name.includes("Mount Shuksan")
    );

    if (adamsPeak) {
      const adamsRoutes = waRoutes.filter((r) => r.area_id === adamsPeak.id);
      console.log(
        `Mount Adams: ${adamsRoutes.length} routes (expected 7)`
      );
      if (adamsRoutes.length !== 7) {
        console.log(
          `  WARNING: Expected 7 routes, found ${adamsRoutes.length}`
        );
      }
      adamsRoutes.forEach((r) => {
        console.log(`  - ${r.name} (${r.grade || "NO GRADE"}) [${r.discipline}]`);
      });
    } else {
      console.log("Mount Adams: NOT FOUND");
    }
    console.log();

    if (shuksanPeak) {
      const shuksanRoutes = waRoutes.filter((r) => r.area_id === shuksanPeak.id);
      console.log(
        `Mount Shuksan: ${shuksanRoutes.length} routes (expected 10)`
      );
      if (shuksanRoutes.length !== 10) {
        console.log(
          `  WARNING: Expected 10 routes, found ${shuksanRoutes.length}`
        );
      }
      shuksanRoutes.forEach((r) => {
        console.log(`  - ${r.name} (${r.grade || "NO GRADE"}) [${r.discipline}]`);
      });
    } else {
      console.log("Mount Shuksan: NOT FOUND");
    }
    console.log();

    // 6. Data completeness check
    console.log("Step 6: DATA COMPLETENESS CHECK");
    console.log("-".repeat(80));

    const dataQualityIssues = {
      missingName: [],
      missingGrade: [],
      missingDiscipline: [],
      missingAreaId: [],
      missingDescription: [],
      nullGrade: [],
      invalidDiscipline: [],
      duplicates: [],
    };

    const VALID_DISCIPLINES = [
      "alpine",
      "mountaineering",
      "rock",
      "ice",
      "mixed",
      "bouldering",
    ];

    // Check for duplicates
    const routeNames = {};
    waRoutes.forEach((r) => {
      const key = `${r.area_id}-${r.name}`;
      if (routeNames[key]) {
        dataQualityIssues.duplicates.push({
          name: r.name,
          areaId: r.area_id,
          ids: [routeNames[key].id, r.id],
        });
      } else {
        routeNames[key] = r;
      }
    });

    waRoutes.forEach((r) => {
      if (!r.name || r.name.trim() === "")
        dataQualityIssues.missingName.push(r);
      if (!r.grade || r.grade.trim() === "")
        dataQualityIssues.missingGrade.push(r);
      if (!r.discipline) dataQualityIssues.missingDiscipline.push(r);
      if (!r.area_id) dataQualityIssues.missingAreaId.push(r);
      if (!r.description || r.description.trim() === "")
        dataQualityIssues.missingDescription.push(r);
      if (r.grade === null || r.grade === undefined)
        dataQualityIssues.nullGrade.push(r);
      if (
        r.discipline &&
        !VALID_DISCIPLINES.includes(r.discipline.toLowerCase())
      ) {
        dataQualityIssues.invalidDiscipline.push(r);
      }
    });

    console.log(`Total routes checked: ${waRoutes.length}`);
    console.log(`Routes missing name: ${dataQualityIssues.missingName.length}`);
    console.log(`Routes missing grade: ${dataQualityIssues.missingGrade.length}`);
    console.log(
      `Routes missing discipline: ${dataQualityIssues.missingDiscipline.length}`
    );
    console.log(
      `Routes missing area_id: ${dataQualityIssues.missingAreaId.length}`
    );
    console.log(
      `Routes missing description: ${dataQualityIssues.missingDescription.length}`
    );
    console.log(`Routes with null grade: ${dataQualityIssues.nullGrade.length}`);
    console.log(
      `Routes with invalid discipline: ${dataQualityIssues.invalidDiscipline.length}`
    );
    console.log(`Duplicate route names: ${dataQualityIssues.duplicates.length}`);

    // Report issues
    if (dataQualityIssues.duplicates.length > 0) {
      console.log("\n*** DUPLICATE ROUTES FOUND ***");
      dataQualityIssues.duplicates.forEach((d) => {
        console.log(
          `  - "${d.name}" in area ${d.areaId} (IDs: ${d.ids.join(", ")})`
        );
      });
    }

    if (dataQualityIssues.missingGrade.length > 0) {
      console.log("\n*** ROUTES MISSING GRADE ***");
      dataQualityIssues.missingGrade.slice(0, 5).forEach((r) => {
        console.log(`  - "${r.name}" (area: ${r.area_id})`);
      });
      if (dataQualityIssues.missingGrade.length > 5) {
        console.log(
          `  ... and ${dataQualityIssues.missingGrade.length - 5} more`
        );
      }
    }

    if (dataQualityIssues.missingDiscipline.length > 0) {
      console.log("\n*** ROUTES MISSING DISCIPLINE ***");
      dataQualityIssues.missingDiscipline.slice(0, 5).forEach((r) => {
        console.log(`  - "${r.name}" (area: ${r.area_id})`);
      });
      if (dataQualityIssues.missingDiscipline.length > 5) {
        console.log(
          `  ... and ${dataQualityIssues.missingDiscipline.length - 5} more`
        );
      }
    }

    if (dataQualityIssues.invalidDiscipline.length > 0) {
      console.log("\n*** ROUTES WITH INVALID DISCIPLINE ***");
      dataQualityIssues.invalidDiscipline.slice(0, 10).forEach((r) => {
        console.log(
          `  - "${r.name}": discipline="${r.discipline}" (area: ${r.area_id})`
        );
      });
      if (dataQualityIssues.invalidDiscipline.length > 10) {
        console.log(
          `  ... and ${dataQualityIssues.invalidDiscipline.length - 10} more`
        );
      }
    }

    console.log();

    // 7. Discipline breakdown
    console.log("Step 7: DISCIPLINE BREAKDOWN");
    console.log("-".repeat(80));
    const disciplines = {};
    waRoutes.forEach((r) => {
      const d = r.discipline || "NONE";
      disciplines[d] = (disciplines[d] || 0) + 1;
    });
    Object.entries(disciplines)
      .sort((a, b) => b[1] - a[1])
      .forEach(([d, count]) => {
        console.log(`${d.padEnd(20)}: ${count}`);
      });
    console.log();

    // 8. Grade distribution
    console.log("Step 8: GRADE DISTRIBUTION (Top 15)");
    console.log("-".repeat(80));
    const grades = {};
    waRoutes.forEach((r) => {
      const g = r.grade || "NONE";
      grades[g] = (grades[g] || 0) + 1;
    });
    const topGrades = Object.entries(grades)
      .sort((a, b) => b[1] - a[1])
      .slice(0, 15);
    topGrades.forEach(([g, count]) => {
      console.log(`${g.padEnd(20)}: ${count}`);
    });
    console.log();

    // 9. Overall health assessment
    console.log("Step 9: OVERALL HEALTH ASSESSMENT");
    console.log("-".repeat(80));

    const totalIssues =
      dataQualityIssues.missingGrade.length +
      dataQualityIssues.missingDiscipline.length +
      dataQualityIssues.missingName.length +
      dataQualityIssues.duplicates.length +
      dataQualityIssues.invalidDiscipline.length;

    const issuePercent = (totalIssues / waRoutes.length) * 100;

    let healthStatus = "EXCELLENT";
    if (issuePercent > 5) healthStatus = "GOOD";
    if (issuePercent > 10) healthStatus = "FAIR";
    if (issuePercent > 20) healthStatus = "POOR";

    console.log(`Total routes: ${waRoutes.length}`);
    console.log(`Total peaks: ${sortedPeaks.length}`);
    console.log(`Issues found: ${totalIssues} (${issuePercent.toFixed(2)}%)`);
    console.log(`Health Status: ${healthStatus}`);
    console.log();

    // 10. Recommendations
    if (totalIssues > 0) {
      console.log("RECOMMENDED FIXES:");
      console.log("-".repeat(80));
      if (dataQualityIssues.missingGrade.length > 0) {
        console.log(
          `1. Add missing grades to ${dataQualityIssues.missingGrade.length} routes`
        );
      }
      if (dataQualityIssues.missingDiscipline.length > 0) {
        console.log(
          `2. Add missing disciplines to ${dataQualityIssues.missingDiscipline.length} routes`
        );
      }
      if (dataQualityIssues.invalidDiscipline.length > 0) {
        console.log(
          `3. Fix invalid disciplines in ${dataQualityIssues.invalidDiscipline.length} routes`
        );
      }
      if (dataQualityIssues.duplicates.length > 0) {
        console.log(
          `4. Remove ${dataQualityIssues.duplicates.length} duplicate route(s)`
        );
      }
    } else {
      console.log("NO ISSUES FOUND - Database is in excellent condition!");
    }

    console.log();
    console.log("=".repeat(80));
    console.log("AUDIT COMPLETE");
    console.log("=".repeat(80));
  } catch (error) {
    console.error("AUDIT ERROR:", error.message);
    process.exit(1);
  }
}

audit();
